crud/common/sharding/bucket_ref_unref.lua

```diff
@@ -0,0 +1,183 @@
+--- module to call vshard.storage.bucket_ref / vshard.storage.bucket_unref
+--- on write requests
+--- there are two modes: safe and fast. on safe mode module
+--- calls vshard.storage.bucket_ref / vshard.storage.bucket_unref
+--- on fast mode it does nothing.
+--- default is fast mode.
+
+--- bucket_ref/bucket_unref must be called in one transaction in order to prevent
+--- safe_mode change during execution.
+
+local vshard = require('vshard')
+local errors = require('errors')
+local rebalance = require('crud.common.rebalance')
+
+local safe_methods
+local fast_methods
+
+local M = {
+    BucketRefError = errors.new_class('bucket_ref_error', {capture_stack = false})
+}
+
+local function make_bucket_ref_err(bucket_id, vshard_ref_err)
+    local err = M.BucketRefError:new(
+        "failed bucket_ref: %s, bucket_id: %s",
+        vshard_ref_err.name,
+        bucket_id
+    )
+    err.bucket_ref_errs = {
+        {
+            bucket_id = bucket_id,
+            vshard_err = vshard_ref_err,
+        }
+    }
+    return err
+end
+
+--- on module initialization safe_mode_status func must be set
+--- it's rebalance.safe_mode_status
+function M.safe_mode_status()
+    error('safe_mode_status not set')
+end
+
+--- Slow bucket_refrw implementation that calls vshard.storage.bucket_refrw.
+--- must be called with bucket_unrefrw in transaction
+function M._bucket_refrw(bucket_id)
+    local ref_ok, vshard_ref_err = vshard.storage.bucket_refrw(bucket_id)
+    if not ref_ok then
+        return false, make_bucket_ref_err(bucket_id, vshard_ref_err)
+    end
+
+    return true
+end
+
+--- Slow bucket_unrefrw implementation that calls vshard.storage.bucket_unrefrw.
+--- must be called with bucket_refrw in transaction
+function M._bucket_unrefrw(bucket_id)
+    return vshard.storage.bucket_unrefrw(bucket_id)
+end
+
+--- Slow bucket_refro implementation that calls vshard.storage.bucket_refro.
+function M._bucket_refro(bucket_id)
+    local ref_ok, vshard_ref_err = vshard.storage.bucket_refro(bucket_id)
+    if not ref_ok then
+        return false, make_bucket_ref_err(bucket_id, vshard_ref_err)
+    end
+
+    return true
+end
+
+--- Slow bucket_unrefro implementation that calls vshard.storage.bucket_unrefro.
+--- must be called in one transaction with bucket_refrw_many
+function M._bucket_unrefro(bucket_id)
+    return vshard.storage.bucket_unrefro(bucket_id)
+end
+
+--- Slow bucket_refrw_many that calls bucket_refrw for every bucket and aggregates errors
+--- @param bucket_ids table<number, boolean>
+function M._bucket_refrw_many(bucket_ids)
+    local bucket_ref_errs = {}
+    local reffed_bucket_ids = {}
+    for bucket_id in pairs(bucket_ids) do
+        local ref_ok, bucket_refrw_err = safe_methods.bucket_refrw(bucket_id)
+        if not ref_ok then
+
+            table.insert(bucket_ref_errs, bucket_refrw_err.bucket_ref_errs[1])
+            goto continue
```

Comment:
Why do we continue on error and then just unref all of them? We should go to unref stright away